### PR TITLE
Generate Lua enums from C++ codebase

### PR DESCRIPTION
**Pull request by ja2-derek**
Proof-of-concept code-gen. See previous discussions: https://github.com/ja2-stracciatella/ja2-stracciatella/pull/1238#discussion_r494956908

This implements a very simple Lua code-generation for enums,using the [magic_enum](https://github.com/Neargye/magic_enum/) library. This does not use the sol2's enum functionality, but instead generate actual Lua code for the enums.

The magic_enum library has its [limitations](https://github.com/Neargye/magic_enum/blob/master/doc/limitations.md), and does not compile on all platforms. The code-generation approach has these advantages:

## Advantages

- Un-intrusive. Other libraries and approaches require us to replace all our plain `enum` with custom macros. 
- `magic_enum`, reflections and code gen logic are not required at runtime. 
- It may help IDE auto-completion of Lua scripts (if there is any)

## Disadvantages and limitations

- Slightly more complex workflow - Lua files have to be generated and committed manually or via CI
- We will have a code-generator to maintain, but it is very simple

---

# Usage

You can build and generate code with:

`./ja2 -enumgen > enums.lua`

It will exit after outputting the code. You can then validate the Lua syntax with:

`lua enums.lua`

Example of a generated enum table:

```lua
--[[
 This file was auto-generated. Do not manually edit.
--]]
Visibility = {
	ANY_VISIBILITY_VALUE = -10,
	HIDDEN_ITEM = -4,
	BURIED = -3,
	HIDDEN_IN_OBJECT = -2,
	INVISIBLE = -1,
	VISIBILITY_0 = 0,
	VISIBLE = 1,
}
```

We will still need to convert some `#define` to enum, and also name all `enum` types.

### CODE REVIEW

**Comment by lynxlynxlynx** [comment]
I wouldn't worry about the compiler requirements — I bet the move to c++17 is more restrictive than that.

**Comment by ja2-derek** [comment]
Actually magic_enum is more restrictive. The `g++` on the Linux CI builder (Ubuntu 18.04) is not supported. It requires Ubuntu 20.04.

File: dependencies/lib-magic_enum/getter/CMakeLists.txt.in [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#  * CMAKE_GENERATOR - which generator to use
+#  * SRC_DIR - where to extract the archive
+#  * BUILD_DIR - where the integrated sources are built
+
+cmake_minimum_required(VERSION 3.1)
+
+project(getter NONE)
+
+include(ExternalProject)
+externalproject_add(get-sol2
```
Comment (lynxlynxlynx) [inline dependencies/lib-magic_enum/getter/CMakeLists.txt.in:14]: Shouldn't this target have a different name?

File: assets/externalized/scripts/enums.lua [hunk 2]
```diff
@@ -1,41 +1,322 @@
...
+	SF_SMOKE_EFFECTS_TEMP_FILE_EXISTS = 1048576,
+	SF_LIGHTING_EFFECTS_TEMP_FILE_EXISTS = 2097152,
+	SF_REVEALED_STATUS_TEMP_FILE_EXISTS = 16777216,
+	SF_DOOR_STATUS_TEMP_FILE_EXISTS = 33554432,
+	SF_ENEMY_PRESERVED_TEMP_FILE_EXISTS = 67108864,
+	SF_CIV_PRESERVED_TEMP_FILE_EXISTS = 134217728,
+	SF_ITEM_TEMP_FILE_EXISTS = 268435456,
+	SF_ROTTING_CORPSE_TEMP_FILE_EXISTS = 536870912,
+	SF_MAP_MODIFICATIONS_TEMP_FILE_EXISTS = 1073741824,
+	SF_DOOR_TABLE_TEMP_FILES_EXISTS = 2147483648,
```
Comment (lynxlynxlynx) [inline assets/externalized/scripts/enums.lua:25]: How does this even work? The limitations you linked say:
> Enum value must be in range [MAGIC_ENUM_RANGE_MIN, MAGIC_ENUM_RANGE_MAX].    By default MAGIC_ENUM_RANGE_MIN = -128, MAGIC_ENUM_RANGE_MAX = 128.

Is this actually about the count of enum values? You're not overriding the range like with some others.

Comment (ja2-derek) [reply assets/externalized/scripts/enums.lua:25]: It uses `magic_enum::flags` so the handling should be different,

File: dependencies/lib-magic_enum/getter/CMakeLists.txt.in (hunk 1 above)
Comment (ja2-derek) [reply dependencies/lib-magic_enum/getter/CMakeLists.txt.in]: Oops. Just fixed.

**Comment by lynxlynxlynx** [comment]
LGTM!

**Pull request merged**